File: examples/rust/src/atomic.rs
#[cfg(test)]
mod tests {
    use std::sync::{
        atomic::{AtomicUsize, Ordering},
        Arc,
    };
    use std::thread;

    #[test]
    fn test_atomic_count_across_threads() {
        const COUNTER_START: usize = 0;
        const NUM_THREADS: usize = 5;

        // create an atomic integer for the counter
        let counter = Arc::new(AtomicUsize::new(COUNTER_START));

        let mut thread_handles = vec![];
        for i in 0..NUM_THREADS {
            let counter = Arc::clone(&counter);

            // spawn a thread that increments the counter
            let thread_handle = thread::spawn(move || {
                // simulate some work
                thread::sleep(std::time::Duration::from_millis(i.try_into().unwrap()));

                // increment the counter atomically
                counter.fetch_add(1, Ordering::Relaxed);

                println!("Thread {} incremented the counter", i);
            });

            thread_handles.push(thread_handle);
        }

        // wait for all threads to finish
        for handle in thread_handles {
            handle.join().unwrap();
        }

        // access the final counter value
        let final_counter = counter.load(Ordering::Relaxed);
        println!("Final counter value: {}", final_counter);

        assert_eq!(final_counter, COUNTER_START + NUM_THREADS);
    }
}

File: examples/c++/tests/atomic.cpp
#include <iostream>
#include <thread>
#include <atomic>
#include <gtest/gtest.h>

std::atomic<int> shared_data(0); // Atomic integer for shared data

const int NUM_THREADS = 5;

void increment_shared_atomic(int thread_id)
{
    // atomically increment the shared data
    shared_data.fetch_add(1, std::memory_order_relaxed);

    std::cout << "Thread " << thread_id
              << " incremented shared_data to " << shared_data.load(std::memory_order_relaxed)
              << std::endl;

    // simulate some work
    std::this_thread::sleep_for(std::chrono::milliseconds(thread_id));
}

TEST(AtomicExample, TestIncrementSharedAtomic)
{
    std::thread thread_handles[NUM_THREADS];

    // Create threads that increment the shared data
    for (int i = 0; i < NUM_THREADS; ++i)
    {
        thread_handles[i] = std::thread(increment_shared_atomic, i);
    }

    // wait for all threads to finish
    for (auto &thread : thread_handles)
    {
        thread.join();
    }

    // print the final value of the shared data
    const auto final_value = shared_data.load(std::memory_order_relaxed);
    std::cout << "Final shared_data value: " << final_value << std::endl;

    EXPECT_EQ(final_value, NUM_THREADS);
}